### ISSUE TITLE
Support Helm’s template + yaml language

### ISSUE THREAD

**Issue by that70schris** [prettier/prettier#6517]
**Prettier 1.18.2**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEUCGBbOSAExg4B0AamgDYCucAzoWgA704C+zIANCBPTAJbTVkoNACcREAO4AFUQkEo0ANwi8AJhxAAjEWjABrODADK9XbygBzZDBFVOACxgYyAdXu941U2DhG5H3kUPAE9kcGpBTnNqOBEYKR0LDDRkADNyGM4AK2oADwAhHX1DI0w4ABlzODSMuGy8o3MLMjgARQoIeBqyTJBTERiRMODMMg16EXMYFzUYe2QADgAGTgmIGJcdejCJmljFas4ARw74BO55EDRqAFooODhVR40ROBPeV4S0JJSkdJ66iAYhheNZbIDqE0Wu1OtU-rVODA0JoZqo5sgAEyInS8MhNADCEAwyTCUGghxAFBiABVkfJ-jFWEA)
```sh
--parser yaml
```

**Input:**
```yaml
name: {{ .Values.app }}
```

**Output:**
```yaml
name: { { .Values.app } }
```

**Expected behavior:**
this messes with helm's ability to interpret yaml files as templates since it does so using double handlebars

**Comment by lydell** [comment]
Isn't that invalid YAML, though? Does this work?

```yaml
name: "{{ .Values.app }}"
```

**Comment by that70schris** [comment]
that might work in certain cases but certainly not all.  Here's some documentation regarding variable usage in Helm: https://v3.helm.sh/docs/topics/chart_template_guide/variables/

essentially although it's technically invalid YAML, Helm uses this syntax to create the actual YAML files that ultimately get used by Kubernetes

**Comment by kachkaev** [comment]
Helm charts _produce_ yaml files, but the original definitions are not strictly yamls themselves. That’s because they contain template directives, which can sometimes be pretty convoluted. The syntax may include ifs, loops and so on, which are impossible to parse as yaml without breaking how normal yamls are turned into ASTs.

Perpahs, someone could develop a Prettier plugin that would handle this superset of yaml? I would not include such parser into Prettier core.

**Comment by lydell** [comment]
I see, thanks.

Still, I think it’s odd that Prettier didn’t throw a parse error on the input, so that’s still something we could improve.

**Comment by ikatyang** [comment]
It's valid YAML (ref: https://yaml.org/spec/1.2/spec.html#id2791704), flow nodes in flow mapping item without `?` or `:` are considered implicit flow mapping pair with flow node as key and null as value. And any flow node could be used as flow pair key, including flow mapping. 

```yaml
{{ .Values.app }}
```

```
flow_mapping
  flow_pair
    key:
      flow_mapping
        flow_pair
          key:
            plain_scalar ".Values.app"
```

**Comment by evilebottnawi** [comment]
@ikatyang so we can't fix it in our original parser?

**Comment by ikatyang** [comment]
It's not a bug, what should we fix?

**Comment by ikatyang** [comment]
You can use `--no-bracket-spacing` to avoid this issue I guess:

**Prettier 1.18.2**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEUCGBbOSAExg4B0AamgDYCucAzoWgA704C+zIANCBPTAJbTVkoNACcREAO4AFUQkEo0ANwi8AJhxAAjEWjABrODADK9XbygBzZADNy1OJwAWMDGQDqj3vGqmwcI3JevIpeAJ7I4NSCnOb2IjBSOhYYaDZ2DiAAVtQAHgBCOvqGRphwADLmcGlk9pzZOUbmFmRwAIoUEPDVtSCmInERoZhkGvQi5jBuajCOyAAcAAycYxD2bjr0EWM0cCKKVZwAjh3widzyIGjUALRQcHCqDxoicMe8L4loyalItjUZ9gwvGQMBEVE41CaLXanSqv3SnBgaE0U1UM2QACZETpeGQmgBhCAYFIRKDQA4gCj2AAqyPkf3srCAA)
```sh
--no-bracket-spacing
--parser yaml
```

**Input:**
```yaml
name: {{ .Values.app }}
```

**Output:**
```yaml
name: {{.Values.app}}

```

**Comment by lydell** [comment]
I googled "yaml parser playground" and picked the first result: https://yaml-online-parser.appspot.com/

That one failed to parse the input, so I thought it wasn’t valid syntax :)

```
ERROR:

while constructing a mapping
  in "<unicode string>", line 1, column 7:
    name: {{ .Values.app }}
          ^
found unacceptable key (unhashable type: 'dict')
  in "<unicode string>", line 1, column 8:
    name: {{ .Values.app }}
```

But I trust @ikatyang more than a random YAML parser on the Internet.

Just now I also tried this one: https://nodeca.github.io/js-yaml/

I manages to parse, but gives some odd output:

```
{ name: { '[object Object]': null } }
```

Maybe that’s expected though, I don’t know yaml.

I’m changing to issue to be a feature request to see how popular this is.

**Comment by that70schris** [comment]
thanks everyone, @ikatyang's workaround works for now.  It'll be interesting to see how popular this is

**Comment by that70schris** [comment]
I've personally started using https://www.pulumi.com/ instead, which makes this a non-issue for me

**Comment by barlock** [comment]
Please re-open this. It's still a bug that others care about.

**Comment by that70schris** [comment]
it seems they do 👍

**Comment by drigz** [comment]
`--no-bracket-spacing` runs into errors in cases like:

```
{{ if .flag }}
key: value
{{ end }}
```
([playground](https://prettier.io/playground/#N4Igxg9gdgLgprEAuExgAICWAzdA6bAGwEMBzdAXwoB0oBrOATyXQDdjCBXOWtBAEyogANCAgAHGJmgBnZKGIAnRRADuABSUI5KYqwiZ+IkACNFxMAxgBlcRcxRSybBxlxRACxgBbQgHUPTHgZOzA4a20gzFYgxmRwGTlRBzdFGHVzUm9iZ1d3EAArGQAPACFzSzgbYm84ABkHOFzCN1Ei4usHUkI4AEVOCHhm1pA7RVT4xhrCY3FFBxg-QxgPZAAOAAZROYg3P3NxeLm4VNYm0QBHAfgMiR0QYhkAWig4OH5340U4K8xvjLI2WG+Tc3kwyBgim4ohkXR6-UGTSQLha+RgxBMS34K2QACZRJDiJhCF0AMIQbxAlBQaDnECcNwAFQxOhRI1Y3AAklAPrBrGB5pIAII86wwRg9YFUIA))

```
key:
  key: {{ aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa }}
```
([playground](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEBrOBPJAdKACPdLPYYPAQ0quptrvocacbwF9XcQAaECABxgBLaAGdkocgCdJEAO4AFKQjEpyANwiCAJtxAAjSeTDoYAZT5HBUAObIAZuQA2IuDwAWMALaOA6m8HwIhZgcKbKAYJqARjI4CJiPFYukjDyhtae5PZOLjwAViIAHgBChsZwZuSecAAyVnDZzq4gBYWmVtaOcACKAK4Q8I25IBaSybEYVY66fJJWMD7aMG7IABwADDyzEC4+hnyxs3DJag08AI798Gn8KiDkIgC0UHBwWm+6knCXgl9p5BkskgHE0eC5PIJkDBJL1miIOl0+gMGsCcs0YOQ9IstMtkAAmHjQ8iCRwdADCEE8mViUGgZxAvRcABVMSoQcM1LCAJJQd6wUxgOYCACCvNMMAwXSGcHYQA))

In some cases, I can imagine improving the results by treating unquoted `{{...}}` as a comment, but this is a hack at best.

In general, it seems impossible to do this without evaluating the Helm templates. For example, Prettier cannot know that it's unsafe to reduce the indentation of `key: value` to 2 spaces:

```
{{ if .flag }}
nested:
  object:
{{ else }}
object:
{{ end }}
    key: value
```

**Comment by gmichaeljaison** [comment]
Until helm chart formatting is supported, auto-formatting can be completely disabled in vscode's settings.json:

```json
"[helm]": {
    "editor.formatOnSave": false
  },
```

**Comment by dfdeagle47** [comment]
> > ```json
> > "[helm]": {
> >     "editor.formatOnSave": false
> >   },
> > ```
> 
> Where do you put that inside settings.json ? just append at the end ?

@leobenkel 

Yes, exactly. For completeness, here's the full procedure:

1. Open up the command palette CTRL/CMD+SHIFT+P.
2. Select "Preferences: Open Settings (JSON)".
3. Add it at the end of the user settings file. It doesn't really matter where you put it.

**Comment by leobenkel** [comment]
Thank you, but I then lose my `yaml` highlight, is it possible to tell vscode to inherit from `yaml` all the settings and override this one ?

**Comment by dfdeagle47** [comment]
> Thank you, but I then lose my `yaml` highlight, is it possible to tell vscode to inherit from `yaml` all the settings and override this one ?

@leobenkel 

Do you have the [Kubernetes VSCode extension](https://marketplace.visualstudio.com/items?itemName=ms-kubernetes-tools.vscode-kubernetes-tools)?

I think it should contain syntax highlighting for Helm (previously contained in https://marketplace.visualstudio.com/items?itemName=technosophos.vscode-helm)

**Comment by leobenkel** [comment]
ah yes ! Thank you !

**Comment by torbendury** [comment]
Does anyone from prettier consider implementing this, or is this issue open FYI?

**Comment by fbartho** [comment]
I believe this is an open issue, and I’m not sure it is solvable. 

Template languages often break the syntax of the co-embedded (interleaved?) language. That means reformatting it isn’t necessarily well defined. (Just like prettier will throw-an-error if you tell it to reformat a TypeScript file with certain kinds of syntax errors)

**Comment by dudicoco** [comment]
Helm templates are definitely not valid yaml files.

Paste this example in and online yaml validator such as https://www.yamllint.com/:
```
{{- range $key, $value := .Values.configMaps -}}

---
apiVersion: v1
kind: ConfigMap
  name: {{ $key }}
  labels:
    {{- include "default.labels" $ | nindent 4 }}
immutable: true
data:
  {{- $data }}

{{- end -}}
```

You will see validation errors.

**Comment by percebus** [comment]
`<rant>`
  My personal preference would be to rename `helm` `*.yaml` files to `*.tmpl.yaml`, like any other pseudo-format (`.tmpl.html`, `.tmpl.json`, etc.), but I'm not sure if `helm` CLI likes that.

  Knowing that `helm/templates` contain YAML templates and not files, helps to simply ignore them via `.prettierignore`.

  Since there is `helm lint`, `helm` should prob ship its own formatter (`helm format`?), like `terraform format` does.
`</rant>`

**Comment by MattJeanes** [comment]
Yeah it should ship its own really, great comparison to Terraform there, could easily be integrated the same way into IDEs

**Comment by lucasfcnunes** [comment]
> Helm templates are definitely not valid yaml files.
> 
> Paste this example in and online yaml validator such as https://www.yamllint.com/:
> 
> ```
> {{- range $key, $value := .Values.configMaps -}}
> 
> ---
> apiVersion: v1
> kind: ConfigMap
>   name: {{ $key }}
>   labels:
>     {{- include "default.labels" $ | nindent 4 }}
> immutable: true
> data:
>   {{- $data }}
> 
> {{- end -}}
> ```
> 
> You will see validation errors.

Yes. [Helm templates are Go templates](https://v3-1-0.helm.sh/docs/howto/charts_tips_and_tricks/#:~:text=Helm%20uses%20Go%20templates%20for%20templating%20your%20resource%20files.%20While%20Go%20ships%20several%20built%2Din%20functions%2C%20we%20have%20added%20many%20others.). What is being asked here is *.yaml.gotmpl formatting.

**Comment by dudicoco** [comment]
> > Helm templates are definitely not valid yaml files.
> > Paste this example in and online yaml validator such as https://www.yamllint.com/:
> > ```
> > {{- range $key, $value := .Values.configMaps -}}
> > 
> > ---
> > apiVersion: v1
> > kind: ConfigMap
> >   name: {{ $key }}
> >   labels:
> >     {{- include "default.labels" $ | nindent 4 }}
> > immutable: true
> > data:
> >   {{- $data }}
> > 
> > {{- end -}}
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > You will see validation errors.
> 
> Yes. [Helm templates are Go templates](https://v3-1-0.helm.sh/docs/howto/charts_tips_and_tricks/#:~:text=Helm%20uses%20Go%20templates%20for%20templating%20your%20resource%20files.%20While%20Go%20ships%20several%20built%2Din%20functions%2C%20we%20have%20added%20many%20others.). What is being asked here is *.yaml.gotmpl formatting.

Not sure if you've read the thread but this is not what's being asked here.

**Comment by lucasfcnunes** [comment]
> Not sure if you've read the thread but this is not what's being asked here.

Helm templates are *yaml.gotmpl. Fixing only one case e.g. `name: { { .Values.app } }` is not going to make prettier "**Support Helm’s template** + yaml language". Attention on "Support Helm’s template" if it's not clear enough.

Read what OP said on https://github.com/prettier/prettier/issues/6517#issuecomment-534770273.
What's cool about gotmpl is that it can template anything html, yaml, etc. Formatting the yaml while maintaining the go template consistency is not trivial (I guess).

**Comment by fbartho** [comment]
@lucasfcnunes Speaking for myself, I came here because I was annoyed that a file with the file extension of`.yaml` was crashing prettier’s yaml parser.

I don’t have files that are tagged with `.gotmpl` indeed, I don’t even know if gotmpl is relevant to my problems.

I want prettier to parse all files tagged with `*.yaml`, this is in a monorepo, so I recognize I need to be annoyed at Helm for co-opting a file format; but that doesn’t make this ticket invalid. Indeed, that ship has sailed, and these files exist in the world — complaining to helm or somebody else belongs in a different ticket.

This ticket is about prettier’s yaml-plugin behavior when it processes real files in the real world.

**Comment by lucasfcnunes** [comment]
> @lucasfcnunes Speaking for myself, I came here because I was annoyed that a file with the file extension of`.yaml` was crashing prettier’s yaml parser.
> 
> I don’t have files that are tagged with `.gotmpl` indeed, I don’t even know if gotmpl is relevant to my problems.
> 
> I want prettier to parse all files tagged with `*.yaml`, this is in a monorepo, so I recognize I need to be annoyed at Helm for co-opting a file format; but that doesn’t make this ticket invalid. Indeed, that ship has sailed, and these files exist in the world — complaining to helm or somebody else belongs in a different ticket on a different repo.
> 
> This ticket is about prettier’s yaml-plugin behavior when it processes real files in the real world.

When did I complain? I want it too.

**Comment by opeologist** [comment]
any updates related here? would like to not have to ignore these files and let prettier do its thing across my repo

**Comment by brainwave** [comment]
Still an issue for me - this feature would be very helpful.

**Comment by mpsanchis** [comment]
I would also appreciate this

**Comment by tduyng** [comment]
As mentioned above, Prettier already supports the "bracketSpacing" feature. If anyone is still looking for a solution, you can easily customize Prettier for specific file extensions in the settings.json file of VS Code.

```json
"[yaml]": {
        "editor.defaultFormatter": "esbenp.prettier-vscode",
        "prettier.tabWidth": 2,
        "prettier.bracketSpacing": false
        //"editor.formatOnSave": false,
    },
    "[yml]": {
        "editor.defaultFormatter": "esbenp.prettier-vscode",
        "prettier.tabWidth": 2,
        "prettier.bracketSpacing": false
        //"editor.formatOnSave": false,
    },
    // Or for those using Kubernetes extensions
    "[helm]": {
        "editor.defaultFormatter": "esbenp.prettier-vscode",
        "prettier.tabWidth": 2,
        "prettier.bracketSpacing": false
        //"editor.formatOnSave": false
    },
```

Should this issue be closed?

**Comment by toniyanchev** [comment]
Still a big issue for me :)

**Comment by ferdinandkeller** [comment]
I'm still having the problem.

**Comment by shane-davidson** [comment]
not sure if a util helper with something like this is the right place to start for this?


```
/**
 * @param {string} text
 * @param {number | false} startIndex
 * @returns {number | false}
 */
function skipInlineGoTemplating(text, startIndex) {
  /* c8 ignore next 3 */
  if (startIndex === false) {
    return false;
  }

  if (text.charAt(startIndex) === "{" && text.charAt(startIndex + 1) === "{") {
    for (let i = startIndex + 2; i < text.length; ++i) {
      if (text.charAt(i) === "}" && text.charAt(i + 1) === "}") {
        return i + 2;
      }
    }
  }
  return startIndex;
}

export default skipInlineGoTemplating;

**Comment by abdukhashimov** [comment]
> You can use `--no-bracket-spacing` to avoid this issue I guess:
> 
> **Prettier 1.18.2** [Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEUCGBbOSAExg4B0AamgDYCucAzoWgA704C+zIANCBPTAJbTVkoNACcREAO4AFUQkEo0ANwi8AJhxAAjEWjABrODADK9XbygBzZADNy1OJwAWMDGQDqj3vGqmwcI3JevIpeAJ7I4NSCnOb2IjBSOhYYaDZ2DiAAVtQAHgBCOvqGRphwADLmcGlk9pzZOUbmFmRwAIoUEPDVtSCmInERoZhkGvQi5jBuajCOyAAcAAycYxD2bjr0EWM0cCKKVZwAjh3widzyIGjUALRQcHCqDxoicMe8L4loyalItjUZ9gwvGQMBEVE41CaLXanSqv3SnBgaE0U1UM2QACZETpeGQmgBhCAYFIRKDQA4gCj2AAqyPkf3srCAA)
> 
> ```shell
> --no-bracket-spacing
> --parser yaml
> ```
> 
> **Input:**
> 
> ```yaml
> name: {{ .Values.app }}
> ```
> 
> **Output:**
> 
> ```yaml
> name: {{.Values.app}}
> ```

`{
  "prettier.bracketSpacing": false
}`

worked on my end

**Comment by jameswilliamknight** [comment]
I found these useful to add to my settings.json in vscode:

```jsonc
    "files.associations": {
        "NOTES.txt": "helm"
    },
    "[helm]": {
        // Prettier was a bit too aggressive with my .yaml files
        "editor.defaultFormatter": "redhat.vscode-yaml",
        // "editor.formatOnSave": false // alternatively, force it to leave you alone
    },
```

Those are the configurations I've collected so far... I'll come back and update this comment if I find anything else useful.

I use the [Kubernetes](https://marketplace.visualstudio.com/items?itemName=ms-kubernetes-tools.vscode-kubernetes-tools) plugin for VS code
Also [redhat.vscode-yaml](https://marketplace.visualstudio.com/items?itemName=redhat.vscode-yaml).

**Comment by ConsoleTVs** [comment]
> ```js
> 
>         "editor.defaultFormatter": "redhat.vscode-yaml",
> ```

I kinda get:

```
FormattingExtension 'YAML' is configured as formatter but it cannot format 'helm-template'-files
```

**Comment by kylemh** [comment]
I don't think it makes sense to solve this issue here.

Kubernetes themselves don't offer a formatter, and RedHat's YAML Formatter also doesn't properly handle their templating language subet.

This feels like it's Kubernetes' responsibility to fix by providing a formatter for us to associate with [helm] files in VS Code.

**Comment by felipecrs** [comment]
But this issue is not about VS Code, is it?

**Comment by kylemh** [comment]
Fair enough. Lost focus on the topic given how it's affecting me and other thread-goers.

Back on focus: Helm really isn't yaml. Its output is. They're in error for not instructing people to avoid using the `.yaml` file suffix. How could prettier possibly know to format yaml like yaml unless it looks like a helm file. Would it even be possible based on AST?

**Comment by felipecrs** [comment]
I think first we would need a parser/formatter for generic go-template. Once we have one, we would need to think on how it can be mixed with the YAML parser/formatter.

**Comment by digitalstudium** [comment]
Just wrote formatter for helm templates: https://github.com/digitalstudium/helmfmt
